src/appleseed.studio/mainwindow/pythonconsole/pythonconsolewidget.cpp

```diff
@@ -121,14 +165,166 @@ void PythonConsoleWidget::slot_execute_all()
     execute(m_input->toPlainText());
 }
 
+void PythonConsoleWidget::execute(const QString& script)
+{
+    PythonInterpreter::instance().execute(script.toStdString().c_str());
+}
+
 void PythonConsoleWidget::slot_clear_output()
 {
     m_output->clear();
 }
 
-void PythonConsoleWidget::execute(const QString& script)
+void PythonConsoleWidget::slot_new_file()
 {
-    PythonInterpreter::instance().execute(script.toStdString().c_str());
+    if (!can_close_file())
+        return;
+
+    close_file();
+}
+
+void PythonConsoleWidget::close_file()
+{
+    m_input->clear();
+    m_opened_file = "";
+    m_is_file_dirty = false;
+}
+
+namespace
+{
+    int show_modified_file_message_box(QWidget* parent)
+    {
+        QMessageBox msgbox(parent);
+        msgbox.setWindowTitle("Save Changes?");
+        msgbox.setIcon(QMessageBox::Question);
+        msgbox.setText("The python script has been modified.");
+        msgbox.setInformativeText("Do you want to save your changes?");
+        msgbox.setStandardButtons(QMessageBox::Save | QMessageBox::Discard | QMessageBox::Cancel);
+        msgbox.setDefaultButton(QMessageBox::Save);
+        return msgbox.exec();
+    }
+}
+
+bool PythonConsoleWidget::can_close_file()
+{
+    // Unmodified file: no problem.
+    if (!m_is_file_dirty)
+        return true;
+
+    // The current file has been modified, ask the user what to do.
+    switch (show_modified_file_message_box(this))
+    {
+        case QMessageBox::Save:
+            slot_save_file();
+            return true;
+
+        case QMessageBox::Discard:
+            return true;
+
+        case QMessageBox::Cancel:
+            return false;
+    }
+
+    assert(!"Should never be reached.");
+    return false;
+}
+
+void PythonConsoleWidget::slot_open_file()
+{
+    if (!can_close_file())
+        return;
+
+    ParamArray settings = PythonInterpreter::instance().get_main_window()->get_settings();
```

Comment:
Modify `get_settings()` to return a mutable reference, and take a mutable reference here; otherwise any changes done by `get_open_filename()` will be lost when exiting this method.